src/drafts/MarkdownViewer/MarkdownViewer.tsx

```diff
@@ -70,25 +64,24 @@ const MarkdownViewer = ({
   onLinkClick,
   openLinksInNewTab = false,
 }: MarkdownViewerProps) => {
-  const outputContainerRef = useRef<HTMLDivElement>(null)
-
-  // Render the HTML into an internal container element so we can modify it before it becomes visible.
-  // Using `unsafeInnerHTML` would require an effect to run after rendering which would cause flicker
-  const [htmlContainer, setHtmlContainer] = useState(() => createRenderedContainer(dangerousRenderedHTML.__html))
-  useEffect(
-    () => setHtmlContainer(createRenderedContainer(dangerousRenderedHTML.__html)),
-    [dangerousRenderedHTML.__html],
-  )
+  // We're using state to store the HTML container because we want the value
+  // to re-run effects when it changes
+  const [htmlContainer, setHtmlContainer] = React.useState<HTMLElement>()
+  const htmlContainerRef = React.useCallback((node: HTMLElement | null) => {
+    if (!node) return
+    setHtmlContainer(node)
+  }, [])
 
   const onChange = useCallback(
     async (value: string) => {
       try {
         await externalOnChange?.(value)
       } catch (error) {
-        setHtmlContainer(createRenderedContainer(dangerousRenderedHTML.__html))
```

Comment:
This reset is important because the state of the DOM of the container updates optimistically. For example:

- User checks a checkbox. The box is now checked
- `externalOnChange` is triggered, allowing the parent to handle the event
- The parent handler errors (ie, API request fails)
- Here, we need to reset the HTML container so the checkbox is unchecked

This is why we were creating the HTML container manually, so we could easily reset it.

Reply:
Updated 👍